kai/pydantic_models.py

```diff
@@ -1,22 +1,74 @@
 import re
 
 from pydantic import BaseModel
+from pygments import lexers
+from pygments.util import ClassNotFound
+
+from kai.kai_logging import KAI_LOG
 
 
 class FileSolutionContent(BaseModel):
     reasoning: str
     updated_file: str
 
 
-# TODO: Sometimes this will fail. Why?
-def parse_file_solution_content(content: str) -> FileSolutionContent:
+def guess_language(code: str, filename: str = None) -> str:
+    try:
+        if filename:
+            lexer = lexers.guess_lexer_for_filename(filename, code)
+            KAI_LOG.debug(f"{filename} classified as {lexer.aliases[0]}")
+        else:
+            lexer = lexers.guess_lexer(code)
+            KAI_LOG.debug(f"Code content classified as {lexer.aliases[0]}\n{code}")
+        return lexer.aliases[0]
+    except ClassNotFound:
+        KAI_LOG.debug(
+            f"Code content for filename {filename} could not be classified\n{code}"
+        )
+        return "unknown"
+
+
+def parse_file_solution_content(language: str, content: str) -> FileSolutionContent:
     reasoning_pattern = r"## Reasoning\s+(.+?)(?=##|$)"
-    updated_file_pattern = r"```(?:\w+)?\s+(.+?)```"
+    code_block_pattern = r"```(?:\w+)?\s+(.+?)```"
 
     reasoning_match = re.search(reasoning_pattern, content, re.DOTALL)
-    updated_file_match = re.search(updated_file_pattern, content, re.DOTALL)
-
     reasoning = reasoning_match.group(1).strip() if reasoning_match else ""
-    updated_file = updated_file_match.group(1).strip() if updated_file_match else ""
+
+    # Only search under the Updated File header if it exists
+    if "## Updated File" in content:
+        updated_file_content = content.split("## Updated File")[1]
+    else:
+        updated_file_content = content
+    code_block_matches = re.findall(code_block_pattern, updated_file_content, re.DOTALL)
+
+    matching_blocks = []
+    for block in code_block_matches:
+        guessed_language = guess_language(block)
+        if language == guessed_language:
+            matching_blocks.append(block)
+
+    if matching_blocks:
+        # If multiple matches default to first
+        updated_file = matching_blocks[0].strip()
+        if len(matching_blocks) > 1:
+            KAI_LOG.debug(
+                f"Multiple matching codeblocks found, defaulting to first {matching_blocks}"
+            )
+        else:
+            KAI_LOG.debug(f"Found single matching codeblock \n{updated_file}")
+    elif code_block_matches:
+        # fallback to first discovered codeblock
+        updated_file = code_block_matches[0]
+        if len(code_block_matches) > 1:
+            KAI_LOG.debug(
+                f"Multiple codeblocks found, defaulting to first {code_block_matches}"
+            )
+        else:
+            KAI_LOG.debug(f"Found single codeblock \n{updated_file}")
+    else:
+        updated_file = ""
+        KAI_LOG.warn("No codeblocks detected in LLM response")
```

Comment:
I'm seeing this more than I think would be expected, I'm worried something is not correct.
```

WARNING - 2024-03-28 13:52:30,093 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:52:30,094 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/utils/StartupListener.java with exception, retrying in 10s


WARNING - 2024-03-28 13:53:31,175 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:53:31,177 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s


WARNING - 2024-03-28 13:56:04,740 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:56:04,742 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s
Error in LLM Response: The LLM did not provide an updated file for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
WARNING - 2024-03-28 13:58:52,980 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:58:52,983 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s
Error in LLM Response: The LLM did not provide an updated file for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
WARNING - 2024-03-28 14:01:47,964 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 14:01:47,965 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s


```

Reply:
it's because mixtral is not providing a complete response